crates/legacy/src/service.rs

```diff
@@ -1235,15 +1778,3188 @@ pub(crate) async fn handle_received_txns<Types: NodeType>(
+            }
+        }
+    }
+
+    /// This test checks that the error `AvailableBlocksError::RequestForAvailableViewThatHasAlreadyBeenDecided`
+    /// is returned when the requested view number has already been garbage
+    /// collected.
+    ///
+    /// To trigger this condition, we initialize the GlobalState with a
+    /// garbage collected view number that is higher than the view that will
+    /// be requested.
+    #[async_std::test]
+    async fn test_get_available_blocks_error_requesting_previous_view_number() {
+        let (bootstrap_sender, _) = async_broadcast::broadcast(10);
+        let (tx_sender, _) = async_broadcast::broadcast(10);
+        let (builder_public_key, builder_private_key) =
+            <BLSPubKey as BuilderSignatureKey>::generated_from_seed_indexed([0; 32], 0);
+        let (leader_public_key, leader_private_key) =
+            <BLSPubKey as SignatureKey>::generated_from_seed_indexed([0; 32], 1);
+        let parent_commit = vid_commitment(&[], 8);
+
+        let state = ProxyGlobalState::<TestTypes>::new(
+            Arc::new(RwLock::new(GlobalState::<TestTypes>::new(
+                bootstrap_sender,
+                tx_sender,
+                parent_commit,
+                ViewNumber::new(0),
+                ViewNumber::new(2),
+                10,
+            ))),
+            (builder_public_key, builder_private_key),
+            Duration::from_millis(100),
+        );
+
+        // leader_private_key
+        let signature = BLSPubKey::sign(&leader_private_key, parent_commit.as_ref()).unwrap();
+
+        // This *should* just time out
+        let result = state
+            .available_blocks_implementation(
+                &vid_commitment(&[], 8),
+                1,
+                leader_public_key,
+                &signature,
+            )
+            .await;
+
+        match result {
+            Err(AvailableBlocksError::RequestForAvailableViewThatHasAlreadyBeenDecided) => {
+                // This is what we expect.
+                // This message *should* indicate that the signature passed
+                // did not match the given public key.
+            }
+            Err(err) => {
+                panic!("Unexpected error: {:?}", err);
+            }
+            Ok(_) => {
+                panic!("Expected an error, but got a result");
+            }
+        }
+    }
+
+    /// This test checks that the error `AvailableBlocksError::GetChannelForMatchingBuilderError`
+    /// is returned when attempting to retrieve a view that is not stored within the state, and
+    /// the highest view is also no longer stored within the state.
+    ///
+    /// To trigger this condition, we initialize the GlobalState with an initial
+    /// state, and then we mutate the state to record the wrong latest state id.
+    /// When interacted with `GlobalState` via `register_builder_state`, and
+    /// `remove_handles`, this error doesn't seem possible immediately possible.
```

Comment:
I'm attempting to state that in order to trigger this condition you must be mutating the internal structure of `GlobalState` directly.  If you instead access it indirectly using `register_builder_state` and the cleanup method `remove_handles` then I don't believe the error is possible to encounter.

In other words.  If you use the `GlobalState` methods as intended instead of the internal structure of the `struct`, this issue does not seem to be able to be triggered.

Reply:
Makes sense!